src/server/game/Movement/MotionMaster.cpp

```diff
@@ -1027,7 +1027,7 @@ void MotionMaster::LaunchMoveSpline(Movement::MoveSplineInit&& init, uint32 id/*
 void MotionMaster::Pop(bool active, bool movementInform)
 {
     MovementGenerator* pointer = *_generators.begin();
-    _generators.erase(pointer);
+    _generators.erase(_generators.begin());
```

Comment:
in case of Chase and Follow, this was deleting both

Reply:
.erase by value will use comparator to determine what to remove (see custom comparator for `_generators`)

Reply:
yeah, the current comparator only checks for type (default, active and passive) and priority, that didn't sound right for this case

Reply:
Are there more suspicious erases like that in code?

Reply:
Umm there are several actually, didnt know that erase by value will use the custom comparator.
This is actually a important issue... IIRC almost all erases rely on "I wan to delete this exact pointer".

What solutions are suitable for this?
- Create a unique identifier for each movement (modifying the container declaration) and erase using that?
- ... ?

Reply:
erase by iterator

Reply:
I'll perform and commit the changes then.

But can you enlighten me about this, please?
* What happens in concurrency scenarios? Example: there are several threads updating the same map context, each thread updates different _parts_, example: map instance script and map entity visitor.
  * Can that actually happen?
  * Can each thread perform actions on the same object at the same time?
  * Is this actually possible? --> Can one thread perform one motionmaster delete at the same time as another thread?
    * This is actually my main worry about iterator erase
    * Is it a mutex or locking variable or deletion queue or any other countermeasure even necessary?

Reply:
All of these questions can be answered with 1 word:
No.

Map is strictly bound to a single thread, there can be no issues like what you described

Reply:
Neat! Ty! 😃

Reply:
TC schedules a map update to the map updater thread pool, waits for all maps to finish the updates, then does some single threaded stuff and restarts. Some events like Object deletion are delayed to avoid race conditions